opm/common/OpmLog/MessageLimiter.hpp

```diff
@@ -0,0 +1,111 @@
+/*
+  Copyright 2016 SINTEF ICT, Applied Mathematics.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_MESSAGELIMITER_HEADER_INCLUDED
+#define OPM_MESSAGELIMITER_HEADER_INCLUDED
+
+#include <string>
+#include <unordered_map>
+
+namespace Opm
+{
+
+
+    /// Handles limiting the number of messages with the same tag.
+    class MessageLimiter
+    {
+    public:
+        /// Used to indicate no message number limit.
+        enum { NoLimit = -1 };
+
+        /// Default constructor, no limit to the number of messages.
+        MessageLimiter()
+            : message_limit_(NoLimit)
+        {
+        }
+
+        /// Construct with given limit to number of messages with the
+        /// same tag.
+        ///
+        /// Negative limits (including NoLimit) are interpreted as
+        /// NoLimit, but the default constructor is the preferred way
+        /// to obtain that behaviour.
+        explicit MessageLimiter(const int message_limit)
+            : message_limit_(message_limit < 0 ? NoLimit : message_limit)
+        {
+        }
+
+        /// The message limit (same for all tags).
+        int messageLimit() const
+        {
+            return message_limit_;
+        }
+
+        /// Used for encounteredMessage() return type (see that
+        /// function).
+        enum class Response
+        {
+            PrintMessage, JustOverLimit, OverLimit
+        };
+
+        /// If a tag is empty, there is no message limit or for that
+        /// tag (count <= limit), respond PrintMessage.
+        /// If (count == limit + 1), respond JustOverLimit.
+        /// If (count > limit + 1), respond OverLimit.
+        Response encounteredMessage(const std::string& tag)
```

Comment:
Better name: `handleTag()` or `includeTag()`? The message itself is not even in scope here!